rfcs/0000-wifi-buddy-interface.md

```diff
@@ -0,0 +1,125 @@
+integrated to serve as a WiFi coprocessor. This device will run MnemOS-based
+firmware, which we are referring to as "WiFi Buddy". See [#121] for details on
+the rationale for choosing the ESP32-C3 as the WiFi device, and for discussion
+of the overall design for ESP32-C3 integration.
+
+The goal of this document is to propose an overall design for the hardware and
+software interfaces between MnemOS on the CPU and MnemOS on the WiFi Buddy.
+
+## Background
+
+### Hardware
+
+The WiFi Buddy hardware device is a [Seeedstudio XIAO] or [Adafruit
+QT Py] ESP32-C3 development board. These are development boards for the ESP32-C3
+which are pinout compatible, so they can be used interchangeably. The ESP32-C3
+board will be treated as a system-on-module (SoM) and integrated with the MnemOS
+Beepy shield using pin headers (or potentially soldered onto the board).
+
+The following pins are broken out by the WiFi Buddy hardware and can be used for
+communication between the CPU and WiFi Buddy:
+
+- One SPI interface (`MOSI`, `MISO`, and `SCK`)
+- One I²C interface (`SDA` and `SCL`)
+- One UART (`TX` and `RX`)
+- Four additional GPIO pins
+
+### Software
+
+Communication between the CPU and the WiFi Buddy will include both a _control
+path_ and a _data path_. Data path communication refers to the actual network
+frames exchanged between the WiFi Buddy and the CPU, while control path
+communication consists of messages that configure the operation of the WiFi
+Buddy.
+
+Potential control path communication may include:
+
+- The CPU asking the WiFi Buddy to scan for WiFi access points
+- The CPU asking the WiFi Buddy to connect to or disconnect from a WiFi access
+  point
+- The ability for the CPU to reset the WiFi Buddy (?)
+- Information about the network interface, such as signal strength
+- Errors and other diagnostic information reported by the WiFi Buddy
+- `trace-proto`-formatted logs from the WiFi buddy (?)
+- Potentially, messages related to Bluetooth Low Energy operation (out of scope
+  for this document)
+
+It may also be desirable for the CPU to eventually be able to update the
+firmware on the WiFI buddy. However, for now, both of the target WiFi Buddy
+devices can be flashed over USB, so the MnemOS CPU being able to update the WiFi
+Buddy firmware is not a high priority at this time.
+
+Data path communication will be performed over the SPI bus, as it's the
+highest-bandwidth link available with this hardware. In [#121], we have
+tentatively concluded that the firmware will use a "dumb WiFi Buddy" design,
+where TCP or any other higher-level network protocols are implemented on the
+CPU, and the CPU and WiFi Buddy communicate by exchanging MAC-layer [802.3]
+Ethernet frames.
+
+## Design Questions
+
+Based on the background information discussed above, we can begin to enumerate
+design questions for the WiFi Buddy interface:
+
+- **Is control path communication in-band or out of band?** If data path frames
```

Comment:
Mentioned in chat, but I'd probably recommend just using SPI unless we come up with a good reason not to.

Reply:
Yeah, we're aligned on this --- I wanted to enumerate both options but the proposed design will be SPI only (and IRQs)